source/wifi/wifi_hal.c

```diff
@@ -6700,10 +6722,175 @@ INT wifi_setRadioStatsEnable(INT radioIndex, BOOL enable)
     return RETURN_ERR;
 }
 
+#ifdef HAL_NETLINK_IMPL
+static int tidStats_callback(struct nl_msg *msg, void *arg) {
+    struct nlattr *tb[NL80211_ATTR_MAX + 1];
+    struct genlmsghdr *gnlh = nlmsg_data(nlmsg_hdr(msg));
+    struct nlattr *sinfo[NL80211_STA_INFO_MAX + 1];
+    struct nlattr *stats_info[NL80211_TID_STATS_MAX + 1],*tidattr;
+    int rem , tid_index = 0;
+
+    wifi_associated_dev_tid_stats_t *out = (wifi_associated_dev_tid_stats_t*)arg;
+    wifi_associated_dev_tid_entry_t *stats_entry;
+
+    static struct nla_policy stats_policy[NL80211_STA_INFO_MAX + 1] = {
+                 [NL80211_STA_INFO_TID_STATS] = { .type = NLA_NESTED },
+    };
+    static struct nla_policy tid_policy[NL80211_TID_STATS_MAX + 1] = {
+                 [NL80211_TID_STATS_TX_MSDU] = { .type = NLA_U64 },
+    };
+
+    nla_parse(tb, NL80211_ATTR_MAX, genlmsg_attrdata(gnlh, 0),
+                  genlmsg_attrlen(gnlh, 0), NULL);
+
+
+    if (!tb[NL80211_ATTR_STA_INFO]) {
+        fprintf(stderr, "station stats missing!\n");
+        return NL_SKIP;
+    }
+
+    if (nla_parse_nested(sinfo, NL80211_STA_INFO_MAX,
+                             tb[NL80211_ATTR_STA_INFO],
+                             stats_policy)) {
+        fprintf(stderr, "failed to parse nested attributes!\n");
+        return NL_SKIP;
+    }
+
+    nla_for_each_nested(tidattr, sinfo[NL80211_STA_INFO_TID_STATS], rem)
+    {
+        stats_entry = &out->tid_array[tid_index];
+
+        stats_entry->tid = tid_index;
+        stats_entry->ac = _tid_ac_index_get[tid_index];
+
+        if(sinfo[NL80211_STA_INFO_TID_STATS])
+        {
+            if(nla_parse_nested(stats_info, NL80211_TID_STATS_MAX,tidattr, tid_policy)) {
+                printf("failed to parse nested stats attributes!");
+                return;
+            }
+        }
+
+        if(stats_info[NL80211_TID_STATS_TX_MSDU])
+            stats_entry->num_msdus = (unsigned long long)nla_get_u64(stats_info[NL80211_TID_STATS_TX_MSDU]);
+
+            tid_index++;
+            printf("all vals are %d , %d , %lld, \n",stats_entry->tid, stats_entry->ac , stats_entry->num_msdus);
+    }
+    //ToDo: sum_time_ms, ewma_time_ms
+
+    return NL_SKIP;
+}
+#endif
+
 INT wifi_getApAssociatedDeviceTidStatsResult(INT radioIndex,  mac_address_t *clientMacAddress, wifi_associated_dev_tid_stats_t *tid_stats,  ULLONG *handle)
 {
-    // TODO Implement me!
-    return RETURN_ERR;
+#ifdef HAL_NETLINK_IMPL
+   Netlink nl;
```

Comment:
Indent should be 4 spaces for this block. Here 3 spaces are given.

Reply:
Noted